### PR TITLE
fsdp support create hybrid-sharded process group for custom backend

### PR DESCRIPTION
**Pull request by medivh-xp**
FSDP creates communication groups for intra-node communication through dist.new_subgroups. Previously, dist.new_subgroups only supported creation based on the number of CUDA devices. However, issue #99706 removed the avaliable-check for CUDA devices, allowing for custom backend create group based on num of custom devices per node.

This PR allows FSDP to explicitly pass device num within the node when creating communication groups for intra-node communication, instead of defaulting to the number of CUDA devices.

### CODE REVIEW

**Comment by medivh-xp** [comment]
The failed tests in the CI seems to be unrelated to this modification:
```bash
bash: ./.ci/pytorch/functorch_doc_push_script.sh: No such file or directory
Error: Process completed with exit code 127.
```

**Review by awgu — APPROVED** [review]
This looks good to me. Thanks!

**Comment by medivh-xp** [comment]
Oh! I made a mistake. The initialization of the process group happens at the beginning of the FSDP init, when the computing device for FSDP has not yet been determined, so it is also impossible to obtain the number of devices...

**Comment by medivh-xp** [comment]
> Oh! I made a mistake. The initialization of the process group happens at the beginning of the FSDP init, when the computing device for FSDP has not yet been determined, so it is also impossible to obtain the number of devices...

Um... Until all the sub-modules are encapsulated by FSDPPolicy, we cannot determine the computing device for FSDP correctly . Additionally, in order for all FSDPs to use the same communication group, we need to complete the initialization of the process before FSDPPolicy takes effect. This seems to be a cycle. However, it seems that we can decide the type of computing device in advance (after marking ignored_params), in order to initialize the device_handle in advance...(｡•́︿•̀｡)? @awgu

**Comment by awgu** [comment]
@medivh-xp I a bit swamped this week. I will try to find some time to understand the issue when I can.

**Comment by medivh-xp** [comment]
> @medivh-xp I a bit swamped this week. I will try to find some time to understand the issue when I can.

Hi, @awgu I moved the initialization of device handle to the beginning of FSDP (after deciding ignored params, as the type of device handle depends on managed params). Wish a glance when you have some free time~

**Review by awgu — COMMENTED** [review]
From an initial look, my main concern is that we may want to do a broader refactor for this. Otherwise, there is duplicate logic between the new `_init_device_handle()` and the existing functions like `_check_single_device_module()` and `_get_compute_device()`.

It seems to me that `_init_device_handle()` moves the logic from `_check_single_device_module()` and `_get_compute_device()` to be earlier. We should de-duplicate those latter ones if we want to go for this.
- Is it true that the `determined_device` in `_init_device_handle()` is always the return value of `_get_compute_device()`?
- Is the single-device-module check in `_init_device_handle()` the same as `_check_single_device_module()`?

**Comment by medivh-xp** [comment]
> 



> From an initial look, my main concern is that we may want to do a broader refactor for this. Otherwise, there is duplicate logic between the new `_init_device_handle()` and the existing functions like `_check_single_device_module()` and `_get_compute_device()`.
> 
> It seems to me that `_init_device_handle()` moves the logic from `_check_single_device_module()` and `_get_compute_device()` to be earlier. We should de-duplicate those latter ones if we want to go for this.
> 
> * Is it true that the `determined_device` in `_init_device_handle()` is always the return value of `_get_compute_device()`?
> * Is the single-device-module check in `_init_device_handle()` the same as `_check_single_device_module()`?

There are still some differences, because auto_wrap (for child modules) depends on the creation of PG, and the creation of PG depends on obtaining the number of devices. Therefore, the type of device needs to be determined before creating PG, so that the device count can be obtained through device_handle.device_count() (currently obtained through torch.cuda.device_count()). Therefore, the current PR hopes to determine the type of device before initializing PG (since the specific device cannot be determined as auto_wrap has not yet been executed).

When _check_single_device_module and _get_compute_device are executed, auto_wrap has already completed, managed parameters have been determined and moved to the correct device, so it can verify that the model is on the same device and determine the specific compute device. However, this cannot be achieved at the beginning of FSDP...

It seems impossible to determine the computing device before auto wrapping, as it is unclear which parameters are being managed and whether they may include meta-type parameters.

File: torch/distributed/fsdp/_init_utils.py [hunk 1]
```diff
@@ -264,6 +265,49 @@ def _init_ignored_module_states(
...
+    if determined_device is None:
+        for param in _get_orig_params(module, ignored_params):
+            if param.device.type in {"cpu", "meta"}:
+                continue
+            if determined_device is None:
+                determined_device = param.device
+            else:
+                if param.device.type != determined_device.type:
+                    raise RuntimeError(
+                        f"FSDP not supports modules on different device type "
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:300]: nit:
```suggestion
                        f"FSDP does not support modules with different device types "
```

**Review by awgu — APPROVED** [review]
This sounds good to me!

**Review by awgu — APPROVED** [review]

**Pull request closed without merging**